util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,190 @@
+
+  private boolean startupCalled = false;
+  private final Stack<StartupStep<CONTEXT>> startedSteps = new Stack<>();
+  private CompletableFuture<CONTEXT> shutdownFuture;
+
+  public StartupProcess(final String processName, final List<StartupStep<CONTEXT>> steps) {
+    this.steps = Collections.unmodifiableList(Objects.requireNonNull(steps));
+
+    this.processName = processName != null ? processName : "Undefined process";
+    logger =
+        LoggerFactory.getLogger(
+            "io.camunda.zeebe.util.startup.StartupProcess(" + this.processName + ")");
+  }
+
+  /**
+   * Executes the startup logic
+   *
+   * @param context the startup context at the start of this step
+   * @return future with startup context at the end of this step
+   */
+  public synchronized CompletableFuture<CONTEXT> startup(final CONTEXT context) {
+    logger.debug("Startup was called with context: " + context);
+    if (startupCalled) {
+      throw new IllegalStateException("startup(...) must only be called once");
+    }
+    startupCalled = true;
+
+    final var startupFuture = new CompletableFuture<CONTEXT>();
+    final var stepsToStart = new ArrayList<>(steps);
+
+    logger.info("Starting startup process");
+    proceedWithStartup(stepsToStart, context, startupFuture);
+
+    return startupFuture;
+  }
+
+  private void proceedWithStartup(
+      final ArrayList<StartupStep<CONTEXT>> stepsToStart,
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> startupFuture) {
+    if (shutdownFuture != null) {
+      logger.info("Aborting startup process because shutdown was called");
+      startupFuture.completeExceptionally(
+          new RuntimeException("Aborting startup process because shutdown was called"));
+    } else if (stepsToStart.isEmpty()) {
+      startupFuture.complete(context);
+      logger.info("Finished startup process");
+    } else {
+      final var stepToStart = stepsToStart.remove(0);
+      startedSteps.push(stepToStart);
+
+      logCurrentStep("Startup", stepToStart);
+      stepToStart
+          .startup(context)
+          .whenComplete(
+              (contextReturnedByStep, error) -> {
+                if (error != null) {
+                  logger.warn("Aborting startup process due to exception.", error);
+                  startupFuture.completeExceptionally(error);
+                } else {
+                  proceedWithStartup(stepsToStart, contextReturnedByStep, startupFuture);
+                }
+              });
+    }
+  }
+
+  /**
+   * Executes the shutdown logic
+   *
+   * @param context the shutdown context at the start of this step
+   * @return future with the shutdown context at the end of this step.
+   */
+  public synchronized CompletableFuture<CONTEXT> shutdown(final CONTEXT context) {
+    logger.debug("Shutdown was called with context: " + context);
+    if (!startupCalled) {
+      throw new IllegalStateException("shutdown(...) can only be called after startup(...)");
```

Comment:
:question: Is this necessary to guard against? There would be no steps to shutdown, so it would be safe to do. Are we guarding against the remote case that we forget to call startup at some point? Improving the `Starting shutdown process` log to include the number of steps required would help diagnose this issue in the small chance that it happens. But I don't think it's so necessary.